wen-restart/src/wen_restart.rs

```diff
@@ -281,18 +297,114 @@ pub(crate) fn find_heaviest_fork(
+    slots: Vec<Slot>,
+    blockstore: Arc<Blockstore>,
+    bank_forks: Arc<RwLock<BankForks>>,
+    root_bank: Arc<Bank>,
+    exit: &Arc<AtomicBool>,
+) -> Result<Hash> {
+    // Find the hash of the heaviest fork, if block hasn't been replayed, replay to get the hash.
+    let heaviest_fork_bankhash = bank_forks
+        .read()
+        .unwrap()
+        .get(heaviest_fork_slot)
+        .map(|bank| bank.hash());
+    if let Some(hash) = heaviest_fork_bankhash {
+        return Ok(hash);
+    }
+
+    let leader_schedule_cache = LeaderScheduleCache::new_from_bank(&root_bank);
+    let replay_tx_thread_pool = rayon::ThreadPoolBuilder::new()
+        .num_threads(1)
+        .thread_name(|i| format!("solReplayTx{i:02}"))
+        .build()
+        .expect("new rayon threadpool");
+    let recyclers = VerifyRecyclers::default();
+    let mut timing = ExecuteTimings::default();
+    let opts = ProcessOptions::default();
+    let mut parent_bank = root_bank;
+    // Grab a big lock because we are the only person touching bankforks now.
+    let mut my_bankforks = bank_forks.write().unwrap();
+    for slot in slots {
+        if exit.load(Ordering::Relaxed) {
+            return Err(WenRestartError::Exiting.into());
+        }
+        let bank = match my_bankforks.get(slot) {
+            Some(cur_bank) => cur_bank,
+            None => {
+                let new_bank = Bank::new_from_parent(
+                    parent_bank.clone(),
+                    &leader_schedule_cache
+                        .slot_leader_at(slot, Some(&parent_bank))
+                        .unwrap(),
+                    slot,
+                );
+                let bank_with_scheduler = my_bankforks.insert_from_ledger(new_bank);
+                let mut progress = ConfirmationProgress::new(parent_bank.last_blockhash());
+                if let Err(e) = process_single_slot(
+                    &blockstore,
+                    &bank_with_scheduler,
+                    &replay_tx_thread_pool,
+                    &opts,
+                    &recyclers,
+                    &mut progress,
+                    None,
+                    None,
+                    None,
+                    None,
+                    &mut timing,
+                ) {
+                    return Err(WenRestartError::BlockNotFrozenAfterReplay(
+                        slot,
+                        Some(e.to_string()),
+                    )
+                    .into());
+                }
+                my_bankforks.get(slot).unwrap()
+            }
+        };
+        if !bank.is_frozen() {
+            return Err(WenRestartError::BlockNotFrozenAfterReplay(slot, None).into());
+        }
+        if bank.parent_slot() != parent_bank.slot() {
+            return Err(WenRestartError::BlockNotLinkedToExpectedParent(
+                slot,
+                Some(bank.parent_slot()),
+                parent_bank.slot(),
+            )
+            .into());
```

Comment:
What do you think if we verified chaining before we start replaying like the previous change? Should be fast and prevent wasting time replaying a bunch of forks only to have it fail on a bad parent

Reply:
Sure. Moved the checks into a separate loop.